README.md
# jss-material-ui
Provide seamless components styling to material-ui

# Introduction
Material-ui 1.0 is a great react library, but its css to jss styling could be enhanced.


# Drawback of the styling provided by material-ui

* doesn't support props
* Styles are applied globally resulting to a lot of warning for the styles not maching the processed component


# The solutions until now to resolve these problems
Some suggest to use styled libraries like styled-component, glamorous, emotion, ... to resove these problems

They resolve it, but two new problems arrive: 

* stylesheet priorisation problem. 
* two css engines on the system which provide more complexity for the app


# A new small styled libray
So I decided to develop a library with the following objective:

* use material-ui styling system, so compatibiity is 100%
* more component oriented
* provide props


# A little sample

## without props


```js
import style from 'jss-material-ui'

// a container style
const ContainerRoot = style(Paper)({
  root: {
    width: '100%',
    marginTop: 3,
    overflowX: 'auto'
  }
})

// container use
<ContainerRoot />
```

The first style is applied to the corresponding styled component.

Its name is generally root. Permits to avoid material-ui warning message.
If you don't have style to pass to the component, you can write 

```js
root: {}
```

## with props

```js
import style from 'jss-material-ui'
import TableCell from '@material-ui/core/TableCell'

const CaloriesCell = style(TableCell)((theme, {calories}) => ({
  root: {
    fontWeight: calories > 300 ? 700 : undefined,
    backgroundColor: calories > 300 ? '#ff0000' : calories < 160 ? '#00FF00' : undefined
  }
}))

// use
<CaloriesCell calories={n.calories} numeric>{n.calories}</CaloriesCell>
```

![Result](./stories/jss.png)




## Custom props

Its also possible to use custom props

```js
style(TableCell, {
     myCustomProps1,
     myCustomProps2
   })
   ((theme, {calories})
```


## Class inherithance

If a child B of a component A has for className :

```js
<A>
   <B className={classes.classB} />
</A>
```

Its possible to write that:

```js
const SA = style(A)((theme, props) => ({
  root: {
    ...
  },
  classB: {
    ...
  }
}))
```

The style specified in the ```classB``` object will be applied to B class component.
Check ```stories/SimpleTableInherit.js```


## ref

To acces the DOM element of the component, use ```sref``` in place of ```ref```

See the storybook Ref example

## more samples

Check ./stories directory for complete samples files.




changelog.md
# Changelog

**2.0.5**
correct dependencies problems on package.json

**2.0.4**
material-ui import optimization
readme.md correction

**2.0.3**
package dependencies change

**2.0.0**
remove duplicated package in package.json
bind with material-ui 3.x

**1.2.1**
storybook Ref modified

**1.2.0**
sref for accessing dom element

**1.1.1**
Remove lock files for installation with latest release

**1.1.0**
Use root attribute for style description

**1.0.8**
Comme back to version 1.0.5

**1.0.5**
Problem with material-ui textfield.
Changed implementation



test/jestsetup.js
import * as TestUtils from 'react-dom/test-utils';
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() })
// Fail tests on any warning
console.error = message => {
  throw new Error(message)
}


test/sample.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import * as TestUtils from 'react-dom/test-utils';
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() })
// Fail tests on any warning
console.error = message => {
    throw new Error(message)
}


it('Test', () => {
    // Render a checkbox with label in the document
    const hello = TestUtils.renderIntoDocument(
        <div>Hello</div>
    );

    const testHello = ReactDOM.findDOMNode(hello);

    // Verify that it's Off by default
    expect(testHello.textContent).toEqual('Hello');
});

src/grid.js
import style from './styled'
import Typography from '@material-ui/core/Typography'

const GridCore = (
  {
    border = false,
    mbetween = false,
    mcenter = false,
    mstart = false,
    scenter = false,
    sstart = false,
    sstrech = false
  }
) => ({
  alignItems: scenter
    ? 'center'
    : sstart
      ? 'flex-start'
      : sstrech
        ? 'stretch'
        : 'center',
  border: border === true ? '1px solid #eee' : 'none',
  display: 'flex',
  flex: '0 1 auto',
  flexDirection: 'row',
  flexWrap: 'wrap',
  justifyContent: mcenter
    ? 'center'
    : mstart
      ? 'flex-start'
      : mbetween
        ? 'space-between'
        : 'space-between',
  marginBottom: 10,
  marginLeft: 20,
  marginRight: 20,
  marginTop: 10,
  paddingBottom: 10,
  paddingLeft: 20,
  paddingRight: 20,
  paddingTop: 10
})

export const HGrid = style('div')((theme, props) => ({
    root: {
        ...GridCore(props),
        flex: 1,
    }
}))

export const H100Grid = style('div')((theme, props) => ({
  root: {
    ...GridCore(props),
    flex: 1,
    width: '100%'
  }
}))

export const VGrid = style('div')((theme, props) => ({
  root: {
    ...GridCore(props),
    flexDirection: 'column',
    flex: 1,
  }
}))

export const V100Grid = style('div')((theme, props) => ({
  root: {
    ...GridCore(props),
    flexDirection: 'column',
    flex: 1,
    width: '100%'
  }
}))

const IGridCore = (
  theme,
  { border = false, center = false, flex = false, start = false, stretch = false }
) => ({
  root: {
    alignSelf: center
      ? 'center'
      : start
        ? 'flex-start'
        : stretch
          ? 'stretch'
          : 'center',
    border: border === true ? '1px solid #eee' : 'none',
    display: 'flex',
    flex: flex !== undefined ? 1 : 0,
    marginBottom: 10,
    marginLeft: 20,
    marginRight: 20,
    marginTop: 10,
    paddingBottom: 10,
    paddingLeft: 20,
    paddingRight: 20,
    paddingTop: 10,
    top: 0
  }
})

export const IGrid = style('div')((theme, props) => IGridCore(theme, props))

export const I100Grid = style('div')((theme, props) => ({
  root: {
    ...IGridCore(theme, props),
    flex: 1,
    width: '100%'
  }
}))

export const Divider = style('hr')({
  root: {
    flex: 1,
    width: '100%',
    border: '1px solid #eee',
    marginBottom: 20
  }
})

export const Space = style('br')({
  root: {
    border: 'none',
    flex: 1,
    width: '100%',
    height: 50
  }
})

export const TitleForm = style(Typography)(
  (theme, { h1 = false, h2 = false, h3 = false }) => ({
    root: {
      align: 'center',
      paragraph: true,
      variant: 'headline',
      paddingBottom: 20,
      component: h1 ? 'h1' : h2 ? 'h2' : h3 ? 'h3' : undefined
    }
  })
)


src/styled.js
import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import withStyles from '@material-ui/core/styles/withStyles';

const styles = (theme, props, style) => {
    return typeof style === 'function'
        ? style(theme, props)
        : style
};

class StyledComponent extends React.Component {
    render() {
        const { classes, className = '', WrappedComponent, sref, ...passThroughProps } = this.props;

        return (
            <WrappedComponent
                classes={classes}
                className={classNames(classes[Object.keys(classes)[0]], className)}
                {...passThroughProps}
                ref={sref ? sref : undefined}
            />
        );
    }
}

StyledComponent.propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
};

const styled = (WrappedComponent, customProps = {}) => {
    return (style, options = {}) => {
        const HOCProps = WrappedComponent => {
            return class _HOCProps extends React.Component {
                FinalComponent =
                        withStyles(theme => styles(theme, this.props, style), options)(StyledComponent)

                render() {
                    return (
                        <this.FinalComponent {...customProps} {...this.props} WrappedComponent={WrappedComponent}/>
                    )
                }
            };
        };
        return HOCProps(WrappedComponent)
    };
};

export default styled

stories/Ref.js
import React from 'react'
import style from '../src/styled'
import { Map } from 'react-leaflet'

const SMap = style(Map)({
    root: {
        margin: 0,
        display: 'flex',
        flex: '1',
        flexDirection: 'row',
        height: '100%',
        width: '100%',
    },
})

class Ref extends React.Component {
    state = {
        mapElement: null
    }
    mapElement = React.createRef()

    componentDidMount = () => {
        this.setState({ mapElement: this.mapElement.current.leafletElement });
    };

    render() {
        return (
            <div>
                <div>
                    <p>Test get ref on Map Leaflet</p>
                    <p>Get Zoom = {this.state.mapElement ? this.state.mapElement.getZoom() : null}</p>
                </div>
                <div>
                    <SMap sref={this.mapElement} zoom={10} />
                </div>
            </div>
        );
    }
};

export default Ref;


stories/SimpleGrid.js
import React from 'react';
import { HGrid, H100Grid, IGrid, TitleForm, VGrid, V100Grid } from '../src/grid';
import TextField from '@material-ui/core/es/TextField/TextField';

class SimpleGrid extends React.Component {
    state = {
        name: 'Cat in the Hat',
    };

    handleChange = event => {
        console.log( event.target.value)
        this.setState({
            [event.target.id]: event.target.value,
        });
    };

    render() {
        return (
            <V100Grid border mCenter>
                <TitleForm>
                    Hello
                </TitleForm>
                <V100Grid border>
                    <H100Grid border mCenter>
                        <VGrid mCenter>
                            <H100Grid mCenter>
                                <p>Hello2</p>
                            </H100Grid>
                        </VGrid>
                        <VGrid mCenter>
                            <TextField
                                id="name"
                                label="Name"
                                value={this.state.name}
                                onChange={this.handleChange}
                                margin="normal"
                            />
                        </VGrid>
                    </H100Grid>
                </V100Grid>
                <V100Grid border>
                    <HGrid border mCenter>
                        <VGrid mCenter>
                            <H100Grid mCenter>
                                <p>Hello2</p>
                            </H100Grid>
                        </VGrid>
                        <VGrid mCenter>
                            <TextField
                                id="name"
                                label="Name"
                                value={this.state.name}
                                onChange={this.handleChange}
                                margin="normal"
                            />
                        </VGrid>
                    </HGrid>
                </V100Grid>
            </V100Grid>
        );
    }
};

export default SimpleGrid;


stories/SimpleTable.js
import React from 'react'
import style from '../src/styled'

import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import _TableCell from '@material-ui/core/TableCell'
import _TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import _Paper from '@material-ui/core/Paper'

const ContainerRoot = style(_Paper)({
  containerRoot: {
    width: '100%',
    marginTop: 3,
    overflowX: 'auto'
  }
})

const TableHead = style(_TableHead)((theme, props) => ({
  tableHead: {
    width: '100%'
  },
}))

const TableCell = style(_TableCell)((theme, {calories}) => ({
  tableCell: {
    fontWeight: calories > 300 ? 700 : undefined,
    backgroundColor: calories > 300 ? '#ff0000' : calories < 160 ? '#00FF00' : undefined
  }
}))

const TableHeadCell = style(_TableCell)({
  tableHeadCell: {
    fontWeight: 700
  }
})

let id = 0

function createData (name, calories, fat, carbs, protein) {
  id += 1
  return {id, name, calories, fat, carbs, protein}
}

const data = [
  createData('Frozen yoghurt', 159, 6.0, 24, 4.0),
  createData('Ice cream sandwich', 237, 9.0, 37, 4.3),
  createData('Eclair', 262, 16.0, 24, 6.0),
  createData('Cupcake', 305, 3.7, 67, 4.3),
  createData('Gingerbread', 356, 16.0, 49, 3.9)
]

const SimpleTable = (props) => {
  return (
    <ContainerRoot>
      <Table>
        <TableHead>
          <TableRow>
            <TableHeadCell>Dessert (100g serving)</TableHeadCell>
            <TableHeadCell numeric>Calories</TableHeadCell>
            <TableHeadCell numeric>Fat (g)</TableHeadCell>
            <TableHeadCell numeric>Carbs (g)</TableHeadCell>
            <TableHeadCell numeric>Protein (g)</TableHeadCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {data.map(n => {
            return (
              <TableRow key={n.id}>
                <TableCell component='th' scope='row'>
                  {n.name}
                </TableCell>
                <TableCell calories={n.calories} numeric>{n.calories}</TableCell>
                <TableCell numeric>{n.fat}</TableCell>
                <TableCell numeric>{n.carbs}</TableCell>
                <TableCell numeric>{n.protein}</TableCell>
              </TableRow>
            )
          })}
        </TableBody>
      </Table>
    </ContainerRoot>
  )
}

export default SimpleTable


stories/SimpleTableInherit.js
import React from 'react';
import style from '../src/styled';

const _ContainerRoot = (props) => {
    const { children, ...rest } = props;

    return (
        <div>
            {children.map((elem, idx) => (React.cloneElement(elem, { key: idx, ...rest })))}
        </div>
    );
};


const ContainerRoot = style(_ContainerRoot)((theme, { backgroundColor, color }) => ({
    root: {
        width: '100%',
        marginTop:
            3,
        overflowX:
            'auto',
    },
    blue: {
        backgroundColor: '#0000ff',
    },

    red: {
        backgroundColor: '#ff0000',
    },

    color: {
        backgroundColor,
        color,
    },
}));


const ClassAA = (props) => {
    const { children, ...rest } = props;

    return (
        <div>
            <p className={props.classes.blue}>Hello blue</p>
            {React.cloneElement(children, { key: 1, ...rest })}
        </div>
    );
};

const ClassA = ({classes, value}) => (<p className={classes.color}>{value}</p>);


const SimpleTableInherit = (props) => {
    return (
        <div>
            <ContainerRoot color={'red'} backgroundColor={'green'}>
                <ClassA value='red/green'/>
                <ClassAA>
                    <ClassA value='red/green'/>
                </ClassAA>
            </ContainerRoot>
            <ContainerRoot color={'blue'} backgroundColor={'#f000f0'}>
                <ClassA value='bleu/violet'/>
                <ClassAA>
                    <ClassA value='bleu/violet'/>
                </ClassAA>
            </ContainerRoot>
        </div>

    );
};

export default SimpleTableInherit;


stories/MixedWithStylesSimpleTable.js
import React from 'react'
import style from '../src/styled'

import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import _TableCell from '@material-ui/core/TableCell'
import _TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import _Paper from '@material-ui/core/Paper'
import withStyles from '@material-ui/core/styles/withStyles';


export const styles = theme => ({
    fat: {
        fontWeight: 700,
        backgroundColor: '#ffff00'
    },
    carbs: {
        fontWeight: 700,
        backgroundColor: '#00ffff'
    },
    protein: {
        fontWeight: 700,
        backgroundColor: '#ff00ff'
    }
});

const ContainerRoot = style(_Paper)({
  root: {
    width: '100%',
    marginTop: 3,
    overflowX: 'auto'
  }
})

const TableHead = style(_TableHead)((theme, props) => ({
  root: {
    width: '100%'
  },
}))

const TableCell = style(_TableCell)((theme, {calories}) => ({
  root: {
    fontWeight: calories > 300 ? 700 : undefined,
    backgroundColor: calories > 300 ? '#ff0000' : calories < 160 ? '#00FF00' : undefined
  }
}))

const TableCellProtein = style(_TableCell)((theme, {protein}) => ({
    root: {
        fontWeight: protein > 4.0 ? 700 : undefined,
        backgroundColor: protein > 5.0 ? '#ff0000' : protein < 4.0 ? '#00FF00' : undefined
    }
}))

const TableHeadCell = style(_TableCell)({
  root: {
    fontWeight: 700
  }
})

let id = 0

function createData (name, calories, fat, carbs, protein) {
  id += 1
  return {id, name, calories, fat, carbs, protein}
}

const data = [
  createData('Frozen yoghurt', 159, 6.0, 24, 4.0),
  createData('Ice cream sandwich', 237, 9.0, 37, 4.3),
  createData('Eclair', 262, 16.0, 24, 6.0),
  createData('Cupcake', 305, 3.7, 67, 4.3),
  createData('Gingerbread', 356, 16.0, 49, 3.9)
]

const SimpleTable = (props) => {
  return (
    <ContainerRoot>
      <Table>
        <TableHead>
          <TableRow>
            <TableHeadCell>Dessert (100g serving)</TableHeadCell>
            <TableHeadCell numeric>Calories</TableHeadCell>
            <TableHeadCell numeric>Fat (g)</TableHeadCell>
            <TableHeadCell numeric>Carbs (g)</TableHeadCell>
            <TableHeadCell numeric>Protein (g)</TableHeadCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {data.map(n => {
            return (
              <TableRow key={n.id}>
                <TableCell component='th' scope='row'>
                  {n.name}
                </TableCell>
                <TableCell calories={n.calories} numeric>{n.calories}</TableCell>
                <_TableCell className={props.classes.fat} numeric>{n.fat}</_TableCell>
                <TableCell className={props.classes.carbs} numeric>{n.carbs}</TableCell>
                <TableCellProtein className={props.classes.protein} protein={n.protein} numeric>{n.protein}</TableCellProtein>
              </TableRow>
            )
          })}
        </TableBody>
      </Table>
    </ContainerRoot>
  )
}

export default withStyles(styles)(SimpleTable)
